init draft

njasm opened the pull request:
This is a work in progress, relates to rust-lang/infra-team#37

So this initial version:
- Listens on ~~127.0.0.1:3000 (will look into listening on all available interfaces if that's what it's wanted)~~ port 3001
- Hyper serves prometheus stats in `/metrics` endpoint, all others will respond a 200 OK
- Github tokens are expected to be in an env var `GITHUB_API_TOKENS` separated by a `,` for more then one. Not having the env var set will panic.
- For each Token/user we are already collection the `rate: limit`, `rate: reset` and `rate: remaining` value. I've also follower @pietroalbini suggestion to have the prometheus label `username=XXX` in each metric. (described on the issue above)
- A very naive data cache for the rate limit collector, right now I spawn a tokio task with a tokio time delay of 120 seconds to query data for the registered tokens.

This lacks alot of stuff like logging/tracing, ~~dockerfile to build and run,~~ how fault-tolerant we wanna be etc. 
But right now I really need community feedback to drive this first PR to completion.

current output for a single collector with a single token provided:
![image](https://user-images.githubusercontent.com/1623137/97792802-32dc0580-1bdb-11eb-9986-d94d381c12b4.png)

Lets see if we can work this one out in the hope of landing a very first initial draft of the monitorbot.

njasm committed:
init draft

njasm committed:
changed listening ip address to 0.0.0.0

njasm committed:
added dotenv crate for easier development workflow. (hidding dev tokens)

njasm committed:
minor refactor dry for http client

njasm committed:
added Dockerfile

emilyalbini commented on Dockerfile:
```diff
@@ -0,0 +1,16 @@
+# Build image
+
+FROM rust:1.47 as build
```
In our Docker images we usually [start from a base Ubuntu image and install Rustup on it](https://github.com/rust-lang/promote-release/blob/master/prod/Dockerfile), but this approach is also fine! What I'd change is building from `stable` instead of pinning 1.47.

njasm replied:
Yes, ok, changed it to `rust:latest` image!

emilyalbini commented on Dockerfile:
```diff
@@ -0,0 +1,16 @@
+FROM rust:1.47 as build
+
+COPY . .
+RUN cargo test --release --all
```
We usually don't execute tests while building the image. Instead, we run tests on CI (using GitHub Actions) and only if they pass we start building the image.

njasm replied:
removed the cargo test run.

emilyalbini commented on Dockerfile:
```diff
@@ -0,0 +1,16 @@
+
+RUN apt-get update && apt-get install -y ca-certificates
+COPY --from=build /target/release/monitorbot /usr/local/bin/
+ENV PORT=3001
```
Our production containers expose the application on port 80.

emilyalbini commented on src/collectors/github_rate_limit.rs:
```diff
@@ -0,0 +1,190 @@
+
+const GH_API_USER_ENDPOINT: &str = "https://api.github.com/user";
+const GH_API_RATE_LIMIT_ENDPOINT: &str = "https://api.github.com/rate_limit";
+const GH_API_TOKENS_ENV_KEY: &str = "GITHUB_API_TOKENS";
```
We usually prefix environment variables with the name of the tool, and I'd make a bit more clear those are the tokens we monitor for rate limits. What about `MONITORBOT_RATE_LIMIT_TOKENS`?

njasm replied:
makes sense, changed to the suggested one!

emilyalbini commented on src/collectors/github_rate_limit.rs:
```diff
@@ -0,0 +1,190 @@
+            Self::RateLimit => client.request(Method::GET, GH_API_RATE_LIMIT_ENDPOINT),
+        };
+
+        rb.header(USER_AGENT, "monitorbot-rust-lang")
```
```suggestion
        rb.header(USER_AGENT, "https://github.com/rust-lang/monitorbot (infra@rust-lang.org)")
```

njasm replied:
done!

emilyalbini commented on src/main.rs:
```diff
@@ -1,3 +1,23 @@
-fn main() {
-    println!("Hello, world!");
+pub mod collectors;
```
I feel like this compiles the `collectors` module twice, one for the library and one for the binary. We should avoid depending on it here, and instead re-rexport just that function from `monitorbot`.

emilyalbini commented on src/main.rs:
```diff
@@ -1,3 +1,23 @@
+#[tokio::main]
+async fn main() -> Result<(), Box<dyn std::error::Error>> {
+    dotenv::dotenv().ok();
+    let addr = "0.0.0.0:3001".parse().unwrap();
```
Note that you're not following the `PORT` environment variable (which it'd be nice to rename to `MONITORBOT_PORT`).

njasm replied:
OK, so we want first try to get a port number from an env var `MONITORBOT_PORT`, if the env var is not set, we fall back to port `80`, correct?

emilyalbini replied:
What I'd do is to keep a non-80 port by default, so that when run the server with `cargo run` locally it doesn't try to bind to the restricted port 80. Then, in the Docker container we set `MONITORBOT_PORT` to 80.

Still, we can fix this after the PR is merged.

emilyalbini reviewed (commented):
Thanks for the PR! The most important parts of it look great!

A couple general points:

* It'd be nice to use as little `pub` as possible, using `pub(crate)` or `pub(super)` instead. The annoying thing about `pub` is it suppresses unused items warnings, because all those items could plausibly be used outside the crate.
* A pattern I started to really like as I do refactoring of our applications is to create a `Config` struct which stores the values of all the environment variables you can set for the application ([example in our release tooling](https://github.com/rust-lang/promote-release/blob/2df758c16e24253257e00a05b44a903b3bf73b04/src/config.rs#L48-L169)). The advantage is, there is a single place to look how to configure the software, instead of wandering through the source code.

njasm committed:
Removed unneeded pub visibility.
Fixed double import of mod collectors
Changed GITHUB_API_TOKENS env var name MONITORBOT_RATE_LIMIT_TOKENS
Grab port from env var MONITORBOT_PORT or fallback to 80
Dockerfile uses now rust:stable to build
Dockerfile removed run command for tests in the build stage

njasm committed:
Removed unneeded pub visibility.
Fixed double import of mod collectors
Changed GITHUB_API_TOKENS env var name MONITORBOT_RATE_LIMIT_TOKENS
Grab port from env var MONITORBOT_PORT or fallback to 80
Dockerfile uses now rust:stable to build
Dockerfile removed run command for tests in the build stage

emilyalbini commented on src/main.rs:
```diff
@@ -1,3 +1,32 @@
-fn main() {
-    println!("Hello, world!");
+use crate::collectors::register_collectors;
```
Shouldn't this be `use monitorbot::collectors::register_collectors;`?

njasm replied:
yes. in fact, that import should be completely removed, and the `fn` I think should be brought into scope in line 3 since we already importing `MetricsProvider` from `monitorbot`. ex: 
```rust
use monitorbot::{MetricProvider, collectors::register_collectors};
```
👍

njasm committed:
resolves https://github.com/rust-lang/monitorbot/pull/1#discussion_r516585660

emilyalbini commented:
Thanks!
